miniapps/elasticity/materials/neohookean.hpp

```diff
@@ -0,0 +1,230 @@
+                 (2.0 / 3.0) * ((i == j) * F[k][l]));
+      });
+   }
+
+   /**
+    * @brief Apply the gradient of the stress.
+    *
+    * @param[in] dudx
+    * @param[in] ddudx
+    * @return
+    */
+   MFEM_HOST_DEVICE tensor<double, dim, dim>
+   action_of_gradient(const tensor<double, dim, dim> &dudx,
+                      const tensor<double, dim, dim> &ddudx) const
+   {
+      if (gradient_type == GradientType::Symbolic)
+      {
+         return action_of_gradient_symbolic(dudx, ddudx);
+      }
+#ifdef MFEM_USE_ENZYME
+      else if (gradient_type == GradientType::EnzymeFwd)
+      {
+         return action_of_gradient_enzyme_fwd(dudx, ddudx);
+      }
+      else if (gradient_type == GradientType::EnzymeRev)
+      {
+         return action_of_gradient_enzyme_rev(dudx, ddudx);
+      }
+#endif
+      else if (gradient_type == GradientType::FiniteDiff)
+      {
+         return action_of_gradient_fd(dudx, ddudx);
+      }
+      else if (gradient_type == GradientType::DualNumbers)
+      {
+         return action_of_gradient_dual(dudx, ddudx);
+      }
+   }
+
+   MFEM_HOST_DEVICE tensor<double, dim, dim>
+   action_of_gradient_dual(const tensor<double, dim, dim> &dudx,
+                           const tensor<double, dim, dim> &ddudx) const
+   {
+      auto sigma = stress(make_tensor<dim, dim>([&](int i, int j)
+      {
+         return mfem::internal::dual<double, double> {dudx[i][j], ddudx[i][j]};
+      }));
+      return make_tensor<dim, dim>(
+      [&](int i, int j) { return sigma[i][j].gradient; });
+   }
+
+#ifdef MFEM_USE_ENZYME
+   MFEM_HOST_DEVICE tensor<double, dim, dim>
+   action_of_gradient_enzyme_fwd(const tensor<double, dim, dim> &dudx,
+                                 const tensor<double, dim, dim> &ddudx) const
+   {
+      tensor<double, dim, dim> sigma{};
+      tensor<double, dim, dim> dsigma{};
+
+      __enzyme_fwddiff<void>(stress_wrapper, enzyme_const, this, enzyme_dup,
+                             &dudx, &ddudx, enzyme_dupnoneed, &sigma, &dsigma);
+      return dsigma;
+   }
+
+   MFEM_HOST_DEVICE tensor<double, dim, dim>
+   action_of_gradient_enzyme_rev(const tensor<double, dim, dim> &dudx,
+                                 const tensor<double, dim, dim> &ddudx) const
+   {
+      tensor<double, dim, dim, dim, dim> gradient{};
+      tensor<double, dim, dim> sigma{};
+      tensor<double, dim, dim> dir{};
+
+      for (int i = 0; i < dim; i++)
+      {
+         for (int j = 0; j < dim; j++)
+         {
+            dir[i][j] = 1;
+            __enzyme_autodiff<void>(stress_wrapper, enzyme_const, this, enzyme_dup,
+                                    &dudx, &gradient[i][j], enzyme_dupnoneed,
+                                    &sigma, &dir);
+            dir[i][j] = 0;
+         }
+      }
+      return ddot(gradient, ddudx);
+   }
+#endif
+
+   MFEM_HOST_DEVICE tensor<double, dim, dim>
+   action_of_gradient_fd(const tensor<double, dim, dim> &dudx,
```

Comment:
`fd` is really close to `fwd`, so I'd call this `action_of_gradient_finite_diff`.